scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -0,0 +1,836 @@
+                       Level.DEBUG);
+        return answer;
+    }
+
+    @Override
+    public <T extends Serializable> T remove(String originator, TaskId taskid, String channel, String key)
+            throws InvalidChannelException, IOException {
+        T answer = ((Channel<T>) getChannel(channel)).remove(key);
+        commitIfNeeded(channel);
+        logWithContext(taskid,
+                       channel,
+                       "Removed key " + QUOTE + key + QUOTE + PREVIOUS_VALUE_WAS + answer,
+                       Level.DEBUG);
+        return answer;
+    }
+
+    @Override
+    public <T extends Serializable> void putAll(String originator, TaskId taskid, String channel,
+            Map<? extends String, ? extends T> m) throws InvalidChannelException, IOException {
+        ((Channel<T>) getChannel(channel)).putAll(m);
+        logWithContext(taskid, channel, "PutAll called, with new entries: " + m, Level.DEBUG);
+        commitIfNeeded(channel);
+    }
+
+    @Override
+    public void clear(String originator, TaskId taskid, String channel) throws InvalidChannelException, IOException {
+        getChannel(channel).clear();
+        logWithContext(taskid, channel, "cleared channel", Level.DEBUG);
+        commitIfNeeded(channel);
+    }
+
+    @Override
+    public Set<String> keySet(String originator, TaskId taskid, String channel) throws InvalidChannelException {
+        return getChannel(channel).keySet();
+    }
+
+    @Override
+    public <T extends Serializable> Collection<T> values(String originator, TaskId taskid, String channel)
+            throws InvalidChannelException {
+        return ((Channel<T>) getChannel(channel)).values();
+    }
+
+    @Override
+    public <T extends Serializable> Set<Map.Entry<String, T>> entrySet(String originator, TaskId taskid, String channel)
+            throws InvalidChannelException {
+        return ((Channel<T>) getChannel(channel)).entrySet();
+    }
+
+    @Override
+    public <T extends Serializable> T getOrDefault(String originator, TaskId taskid, String channel, String key,
+            T defaultValue) throws InvalidChannelException {
+        return ((Channel<T>) getChannel(channel)).getOrDefault(key, defaultValue);
+    }
+
+    @Override
+    public void forEach(String originator, TaskId taskid, String channel, String action)
+            throws CompilationException, ClosureEvaluationException, InvalidChannelException {
+        try {
+            Channel<Serializable> chosenChannel = getChannel(channel);
+            chosenChannel.forEach(evaluateClosure(action, BiConsumer.class));
+            logWithContext(taskid, channel, "Run forEach, new content is: " + chosenChannel, Level.DEBUG);
+        } catch (InvalidChannelException | CompilationException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new ClosureEvaluationException(EXCEPTION_WHEN_EVALUATING_CLOSURE + StackTraceUtil.getStackTrace(e));
```

Comment:
That's what I did in the beginning, unfortunately, exceptions thrown by groovy are often not Serializable, and thus cannot be transfered by ProActive to the user, and thus actually lost.

Converting the exception to string is the only workaround I found.

Reply:
Wouldn't it be possible to catch the `NotSerializableException` here and then use the work around. 
So that we could be able to chain the initial exception for the exceptions that are serializable.

Reply:
Unfortunately, the NotSerializableException is triggered by ProActive when serializing back the exception to the caller. It is nowhere inside the method implementation that it is triggered.